### PR TITLE
Remove song from playlist

### PR DESCRIPTION
**Pull request by ejohnson44**
Allows songs to be removed from the playlist. Also added in temporary toast messages for adding to and removing from the playlist

### CODE REVIEW

File: SoundStream/src/com/lastcrusade/soundstream/components/MusicLibraryFragment.java [hunk 1]
```diff
@@ -137,7 +150,7 @@ protected PlaylistService getPlaylistService() {
      * Inner class extends MusicListAdapter to add the add to playlist image button, and click listener
      * 
      */
-    private class MusicAdapter extends MusicListAdapter {
+    private class MusicAdapter extends MusicListAdapter<SongMetadata> {
```
Comment (reidbaker) [inline SoundStream/src/com/lastcrusade/soundstream/components/MusicLibraryFragment.java:153]: Why was this added?

File: SoundStream/src/com/lastcrusade/soundstream/components/PlaylistFragment.java [hunk 2]
```diff
@@ -92,7 +96,19 @@ private void registerReceivers() {
...
-        }).register(this.getActivity());
+        })
+        .addAction(PlaylistService.ACTION_SONG_REMOVED, new IBroadcastActionHandler() {
+            
+            @Override
+            public void onReceiveAction(Context context, Intent intent) {
+                SongMetadata entry = intent.getParcelableExtra(PlaylistService.EXTRA_SONG);
+                //for now we are just toasting, but eventually this might change to something that 
+                //allows the user to undo the action
+                Toaster.iToast(getActivity(), "\"" + entry.getTitle() + "\" has been removed.");
```
Comment (reidbaker) [inline SoundStream/src/com/lastcrusade/soundstream/components/PlaylistFragment.java]: Since we are planning on always telling the user in some way they removed a song can put this in strings.xml?
Maybe "Removed: " as the string

File: SoundStream/src/com/lastcrusade/soundstream/components/PlaylistFragment.java [hunk 3]
```diff
@@ -139,7 +155,29 @@ public View getView(int position, View convertView, ViewGroup parent) {
             } else {
                 element.setBackgroundColor(getResources().getColor(com.actionbarsherlock.R.color.abs__background_holo_light));
             }
+            
+            ImageButton delete = (ImageButton)element.findViewById(R.id.btn_remove_from_playlist);
+            delete.setOnClickListener(new DeleteSongListener(entry));
+            delete.setVisibility(View.VISIBLE);
+
```
Comment (reidbaker) [inline SoundStream/src/com/lastcrusade/soundstream/components/PlaylistFragment.java:162]: Lot of white space

File: SoundStream/src/com/lastcrusade/soundstream/components/PlaylistFragment.java [hunk 4]
```diff
@@ -139,7 +155,29 @@ public View getView(int position, View convertView, ViewGroup parent) {
...
+        }
+        
+        private class DeleteSongListener implements OnClickListener{
+            private PlaylistEntry entry;
+            public DeleteSongListener(PlaylistEntry entry){
+                this.entry = entry;
+            }
+            @Override
+            public void onClick(View v) {
+                getPlaylistService().removeSong(entry);
```
Comment (reidbaker) [inline SoundStream/src/com/lastcrusade/soundstream/components/PlaylistFragment.java:176]: Have we tested this with a song that is currently being played

File: SoundStream/src/com/lastcrusade/soundstream/components/MusicLibraryFragment.java (hunk 1 above)
Comment (ejohnson44) [inline SoundStream/src/com/lastcrusade/soundstream/components/MusicLibraryFragment.java:153]: Music List Adapter got changed to take in generics - Eclipse was telling me this should be there so I added it in.

File: SoundStream/src/com/lastcrusade/soundstream/components/PlaylistFragment.java (hunk 2 above)
Comment (ejohnson44) [inline SoundStream/src/com/lastcrusade/soundstream/components/PlaylistFragment.java]: The reason I didn't add that in as a string was because I wasn't sure if we were going to keep it around as a word or if we wanted to change to some purely visual feedback. But I can go ahead and put that in, and if we decide on something else later I can just remove it.

File: SoundStream/src/com/lastcrusade/soundstream/model/Playlist.java [hunk 5]
```diff
@@ -37,7 +37,10 @@ public SongMetadata remove(SongMetadata meta) {
         SongMetadata removeMeta = meta;
         boolean removed = musicList.remove(meta);
         if (!removed) {
-            removeMeta = null;
+            removed = playedList.remove(meta);
+            if(!removed){
```
Comment (reidbaker) [inline SoundStream/src/com/lastcrusade/soundstream/model/Playlist.java]: I think this code could be clearer. It used to say remove from the musicList now it can remove from either.

``` java
boolean success;
SongMetadata removeMeta = meta;
if(musicList.contains(meta)) {
    success = musicList.remove(meta);
} else if(playedList.contains(meta){
    success = playedList.remove(meta)
} else {
    success = false; 
    Log.wtf(TAG, "Asked to remove unknown object);
}

if(!success){
    removeMeta = null;
}
```

File: SoundStream/src/com/lastcrusade/soundstream/components/PlaylistFragment.java (hunk 4 above)
Comment (ejohnson44) [inline SoundStream/src/com/lastcrusade/soundstream/components/PlaylistFragment.java:176]: It removes but does not stop playing - I'll work on that.

File: SoundStream/src/com/lastcrusade/soundstream/service/PlaylistService.java [hunk 6]
```diff
@@ -310,11 +315,35 @@ public void addSong(PlaylistEntry entry) {
         if (isLocalPlayer()) {
             mDataManager.addToLoadQueue(entry);
         }
+        new BroadcastIntent(ACTION_SONG_ADDED)
+            .putExtra(EXTRA_SONG, entry)
+            .send(this);
+        // send an intent to the fragments that the playlist is updated
         new BroadcastIntent(ACTION_PLAYLIST_UPDATED).send(this);
+        //send a message to the network that the playlist is updated
+        ((CustomApp)this.getApplication()).getMessagingService().sendPlaylistMessage(mPlaylist.getSongsToPlay());
```
Comment (reidbaker) [inline SoundStream/src/com/lastcrusade/soundstream/service/PlaylistService.java:324]: This is a long line

**Comment by reidbaker** [comment]
@ejohnson44 and @theJenix now that we have an Add song intent. Does that mean we should not call the playlist service from the MusicLibrary fragment.

File: SoundStream/src/com/lastcrusade/soundstream/service/PlaylistService.java (hunk 6 above)
Comment (theJenix) [inline SoundStream/src/com/lastcrusade/soundstream/service/PlaylistService.java:324]: This is going to be changed when I check in the MusicLibraryServiceTest fix.  Not a showstopper for now.
